ARCHITECTURE.md

```diff
@@ -0,0 +1,221 @@
+ - `"ANY"` (#0 - #5 or #7)
+ - `"OTHER"` (N/A)
+
+`"OTHER"` means another type defined in the same document, and is used as a pointer to that type definition.
+The CDDL code that can give rise to these are described in the [README](README.md).
+
+`self.value` means different things for different values of `self.type`.
+E.g. for `"UINT"`, `self.value` has the value dictated by the type, or `None` if different values are allowed, so in the following example, `Foo` will have a `self.value` of 5, and `Bar` will have `None`.
+
+```cddl
+Foo = 5
+Bar = uint
+```
+
+For container types, i.e `"LIST"`, `"MAP"`, `"GROUP"`, and `"UNION"`, `self.value` contains a list of their contents.
+For `"OTHER"`, `self.value` is a string with the name of the type it refers to.
+The following example shows use of both container and `"OTHER"` types.
+
+```cddl
+Foo = uint
+Bar = [*Foo]
+```
+
+This will spawn 3 CddlParser objects:
+
+1. `Foo`, which has `self.type = "UINT"` and `self.value = None`
+2. An anonymous child of Bar, which has `self.type = "OTHER"`, and `self.value = "Foo"`
+3. Bar, which has `self.type = "LIST"`, and `self.value` is a python `list` containing the above object.
+
+CDDL supports many other constraints on the types, and these all have member variables in CddlParser, e.g. `self.minQ` and `self.maxQ` which are the minimum and maximum quantity/repetitions of this type.
+
+Children of `"MAP"` objects come in key/value pairs.
+These are represented such that the values will be children of the `"MAP"` object, and the keys can be found as `self.key` in these children.
+
+There is a member called `self.my_types`, which is a dict of all the types defined in the CDDL file.
+The elements are on the form `<name>: <CddlParser object>`.
+`"OTHER"` objects will look into `self.my_types[self.value]` to find its own definition.
+
+The actual parsing of the CDDL happens with regular expressions.
+There are one or more expressions for each base type.
+The expressions consume a number of characters from the input string, and also capture a substring to use as the value of the type.
+For container types, the regex will match the whole type, and then recursively parse the children within the matched string.
+
+CddlXcoder
+----------
+
+CddlXcoder inherits from CddlParser, and provides common functionality for PythonClass and CodeGenerator.
+
+Most of the functionality falls into one of two categories:
+
+- Common names for members in generated code. A single type possibly need multiple member variables in the generated code to describe it, like
+   - the value
+   - the key associated with this value
+   - the number of times it repeats
+- Condition functions that make inferences about the type based on the member variables in CddlParser, like:
+   - key_var_condition(): Whether it needs a key member
+   - is_unambiguous(): Whether the type is completely specified, i.e. whether we know beforehand exactly how the encoding will look (e.g. `Foo = 5`).
+
+PythonClass 
```

Comment:
Given:

> PythonClass...allows converting data between a number of different representations like CBOR/YAML/JSON strings, but also internal Python representations.

I don't think this class name accurately represents its purpose, and instead comes across like a custom redefinition of the built-in `object`.

Perhaps something like `DataTranslator`, `DataConverter`, `TypeConverter`, `CborManipulator`, etc. would help make it more self-explanatory when seeing it used.

Reply:
Good point. The name has been nagging me, but I never got around to changing it.
I think I'll go with `DataTranslator`. Thanks.